File: app/src/main/java/com/syroniko/casseteapp/ChatAndMessages/MessagesFragment.java
package com.syroniko.casseteapp.ChatAndMessages;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.google.firebase.firestore.QuerySnapshot;
import com.syroniko.casseteapp.MainClasses.MainActivity;
import com.syroniko.casseteapp.MainClasses.User;
import com.syroniko.casseteapp.R;
import com.syroniko.casseteapp.room.UserAndTime;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import static com.syroniko.casseteapp.utils.ArrayUtilsKt.sortUserList;

public class MessagesFragment extends Fragment {

    private RecyclerView recyclerView;
    private static ArrayList<UserAndTime> list;
    private FriendChatListAdapter adapter;
    private EditText searchEditText;
    private String uid;
    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState)
    {
        View view =inflater.inflate(R.layout.fragment_messages,container,false);
        final FirebaseFirestore db = FirebaseFirestore.getInstance();
        final Fragment fragment = this;

        recyclerView=view.findViewById(R.id.recycler_of_friends_message_fragment);
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        searchEditText=view.findViewById(R.id.search_users_edit_text);
        searchEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                searchUsers(s.toString().toLowerCase());
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
        list=new ArrayList<>();
        uid = ((MainActivity) getActivity()).getUid();

        DocumentReference docRef = db.collection("users").document(uid);

        docRef.get().addOnCompleteListener(new OnCompleteListener<DocumentSnapshot>() {
            @Override
            public void onComplete(@NonNull Task<DocumentSnapshot> task) {
                if (task.isSuccessful()) {
                    DocumentSnapshot document = task.getResult();

                    if (document.exists()) {
                       final HashMap<String, Long> friends= (HashMap<String, Long>) document.get("friends");

//                       for(int i =0;i<friends.size();i++){
                        for (final Map.Entry<String, Long> entry : friends.entrySet()){
                           DocumentReference friendRef = db.collection("users").document(entry.getKey());
                           friendRef.get().addOnCompleteListener(new OnCompleteListener<DocumentSnapshot>() {
                               @Override
                               public void onComplete(@NonNull Task<DocumentSnapshot> task) {
                                   if (task.isSuccessful()) {
                                       DocumentSnapshot friendDocument = task.getResult();

                                       if (friendDocument.exists()) {
                                           User newUser = friendDocument.toObject(User.class);

                                           list.add(new UserAndTime(newUser, entry.getValue()));
                                           list = sortUserList(list);
                                           Log.v("zazaza", String.valueOf(list.size()));
                                       }
                                   }

                                   adapter = new FriendChatListAdapter(getContext(), list);
                                   recyclerView.setAdapter(adapter);
                                   adapter.notifyDataSetChanged();
                               }
                           });
                        }
                        ChatUtilsKt.insertInLocalDb(fragment, list);

                    } else {
                        Log.d("TAG", "No such document");
                    }


                } else {
                    Log.d("TAG", "get failed with ", task.getException());

                }
            }
        });


        return view;
    }

    @Override
    public void onStart() {
        super.onStart();

        if (adapter == null || list == null){
            return;
        }
        sortUserList(list);
        Log.d(MessagesFragment.class.getSimpleName(), list.toString());
        adapter.notifyDataSetChanged();
        Toast.makeText(getContext(), "AHA", Toast.LENGTH_SHORT).show();
    }

    private void searchUsers(String s) {
        final FirebaseUser firebaseUser= FirebaseAuth.getInstance().getCurrentUser();
        FirebaseFirestore db = FirebaseFirestore.getInstance();

        CollectionReference colRef = db.collection("users");
        colRef.orderBy("name").startAt(s).endAt(s+"\uf8ff")
                .addSnapshotListener(new EventListener<QuerySnapshot>() {
                    @Override
                    public void onEvent(@javax.annotation.Nullable QuerySnapshot value,
                                        @javax.annotation.Nullable FirebaseFirestoreException e) {
                        Log.v("AKA",String.valueOf(value.getDocuments().size()));

                        if (e != null) {
                            return;
                        }


                        list.clear();
                        for (QueryDocumentSnapshot doc : value) {
                            String name=  doc.getString("name");
                            name=name.toLowerCase();
                            String isOnline=doc.getString("status");
                            User user=new User(name,doc.getString("email"),isOnline,doc.getLong("receivedLastCassetteAt").longValue(),null,doc.getString("uid"),doc.getString("country"),
                                    new ArrayList<String>(),new HashMap<String, Long>(), new ArrayList<String>(), 0, 0, new ArrayList<String>(), "", new ArrayList<String>(), new ArrayList<String>());
                            if(!user.getUid().equals(firebaseUser.getUid())){
                                list.add(new UserAndTime(user, 0));
                            }
                        }
                        adapter=new FriendChatListAdapter(getContext(),list);
                        recyclerView.setAdapter(adapter);
                    }
                });
        }


    }

File: app/src/main/java/com/syroniko/casseteapp/chatAndMessages/MessagesFragment.kt
package com.syroniko.casseteapp.chatAndMessages

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.syroniko.casseteapp.mainClasses.MainViewModel
import com.syroniko.casseteapp.R
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject


@AndroidEntryPoint
class MessagesFragment : Fragment() {

    private val viewModel by activityViewModels<MainViewModel>()
    @Inject lateinit var chatsAdapter: FriendChatListAdapter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_messages, container, false)

        val recyclerView: RecyclerView = view.findViewById(R.id.recycler_of_friends_message_fragment)
        val searchEditText: EditText = view.findViewById(R.id.search_users_edit_text)

        if (viewModel.chats.value != null) {
            chatsAdapter.displayedChats = viewModel.chats.value!!
        }

        recyclerView.apply {
            setHasFixedSize(true)
            layoutManager = LinearLayoutManager(context)
            adapter = chatsAdapter
        }

        viewModel.chats.observe(viewLifecycleOwner, Observer { displayedChatsList ->
//            old code to update the chat, keep it here until we ensure that the new one works
//            chatsAdapter.displayedChats = displayedChatsList ?: return@Observer

            //we use "<" instead of "!=" in order to be able to work in cases where a new contact or
            //a very old contact (not loaded without scroll down) send a message and there are more
            //documents in the displayedChatList
            //IMP: currently this isn't implemented, but it's written like that to make it in the future
            if (displayedChatsList.size < chatsAdapter.displayedChats.size){
                return@Observer
            }

            //used to smoothly update the positions of the items in the adapter (see below)
            val positionsMap = mutableMapOf<Int, String>()

            //for each new chat and each chat in the adapter check if they have the same chatId but
            //they are different (this means that this chat has been updated
            //if yes, check if they have different timestamps (this means that a new message was
            //sent, if they do update the positions map
            //finally change the item in the adapter and notify the adapter
            //important: if we use notifyDataSetChange the update animation is very bad
            displayedChatsList.map { displayedChat ->
                for (i in 0 until chatsAdapter.displayedChats.size){
                    val adapterChatI = chatsAdapter.displayedChats[i]

                    if (displayedChat.chatId == adapterChatI.chatId && displayedChat != adapterChatI){
                        if (displayedChat.timestamp != adapterChatI.timestamp) {
                            //we temporarily set the chatId as value to be able to identify the
                            //chat later (that's also why we use String as the value's type)
                            //however the value of the map is not intended for this
                            positionsMap[i] = adapterChatI.chatId
                        }
                        chatsAdapter.displayedChats[i] = displayedChat
                        chatsAdapter.notifyItemChanged(i)
                    }
                }
            }

            //sort the new chats (newest goes first)
            chatsAdapter.displayedChats.sortByDescending { it.timestamp }

            //for each of the newly updated chats (in the adapter) check if their chatId exists as
            //value of the map
            //if yes, replace that value with the current position of the chat, this way the map
            //will show the change of the positions like that:
            //key = oldPosition -> value = newPosition
            var i = 0
            chatsAdapter.displayedChats.map { chat ->
                positionsMap.map { entry ->
                    if (chat.chatId == entry.value){
                        positionsMap[entry.key] = i.toString()
                        i++
                    }
                }
            }

            //for every entry of the positions map notify the adapter that the item is moved
            //from the old position (key) to the new position (value)
            //once again if we just use notifyDataSetChange the update animation will be a problem
            positionsMap.map { entry ->
                chatsAdapter.notifyItemMoved(entry.key, entry.value.toInt())
            }

        })

        searchEditText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                chatsAdapter.sortListFromSearch(s.toString())
            }
            override fun afterTextChanged(s: Editable) {}
        })

        return view
    }

}